### PR TITLE
*: remove CancelFunc return for Watch. Use Cancel for watch.

### PR DESCRIPTION
**Pull request by gyuho**
@xiang90 @heyitsanthony

### CODE REVIEW

File: storage/watcher.go [hunk 1]
```diff
@@ -92,12 +90,49 @@ func (ws *watchStream) Chan() <-chan WatchResponse {
 }
 
 func (ws *watchStream) Cancel(id int64) error {
```
Comment (gyuho) [inline storage/watcher.go]: This is inefficient. Will change it to keep `*watcher` to cancel function map.

**Comment by gyuho** [comment]
@xiang90 @heyitsanthony Please take a look again. Changed `CancelFunc` to internal type,
and remove `CancelFunc` return in `Watch` method.

File: storage/watchable_store_bench_test.go [hunk 2]
```diff
@@ -60,11 +60,11 @@ func BenchmarkWatchableStoreUnsyncedCancel(b *testing.B) {
...
 	watcherN := k * benchSampleN
 
-	cancels := make([]CancelFunc, watcherN)
+	watchIDs := make([]int64, watcherN)
 	for i := 0; i < watcherN; i++ {
 		// non-0 value to keep watchers in unsynced
-		_, cancel := w.Watch(testKey, true, 1)
-		cancels[i] = cancel
+		id := w.Watch(testKey, true, 1)
+		watchIDs[i] = id
```
Comment (heyitsanthony) [inline storage/watchable_store_bench_test.go]: save a line with `watchIDs[i] = w.Watch(...)`

File: storage/watchable_store_bench_test.go [hunk 3]
```diff
@@ -76,7 +76,10 @@ func BenchmarkWatchableStoreUnsyncedCancel(b *testing.B) {
 
 	// cancel N watchers
 	for _, idx := range ix[:benchSampleN] {
-		cancels[idx]()
+		id := watchIDs[idx]
+		if err := w.Cancel(id); err != nil {
```
Comment (heyitsanthony) [inline storage/watchable_store_bench_test.go]: save a line, `w.Cancel(watchIDs[idx])`

File: storage/watchable_store_bench_test.go [hunk 4]
```diff
@@ -97,11 +100,11 @@ func BenchmarkWatchableStoreSyncedCancel(b *testing.B) {
...
 	const watcherN = 1000000
 
-	cancels := make([]CancelFunc, watcherN)
+	watchIDs := make([]int64, watcherN)
 	for i := 0; i < watcherN; i++ {
 		// 0 for startRev to keep watchers in synced
-		_, cancel := w.Watch(testKey, true, 0)
-		cancels[i] = cancel
+		id := w.Watch(testKey, true, 0)
+		watchIDs[i] = id
```
Comment (heyitsanthony) [inline storage/watchable_store_bench_test.go]: `watchIDs[i] = w.Watch(...)`

File: storage/watchable_store_bench_test.go [hunk 5]
```diff
@@ -112,6 +115,9 @@ func BenchmarkWatchableStoreSyncedCancel(b *testing.B) {
 	b.ReportAllocs()
 
 	for _, idx := range ix {
-		cancels[idx]()
+		id := watchIDs[idx]
+		if err := w.Cancel(id); err != nil {
```
Comment (heyitsanthony) [inline storage/watchable_store_bench_test.go]: `w.Cancel(watchIDs[idx])`

File: storage/watchable_store_test.go [hunk 6]
```diff
@@ -96,17 +98,18 @@ func TestCancelUnsynced(t *testing.T) {
...
-	// synced and unsynced
-	cancels := make([]CancelFunc, watcherN)
+	// create watcherN of watch ids to cancel
+	watchIDs := make([]int64, watcherN)
 	for i := 0; i < watcherN; i++ {
 		// use 1 to keep watchers in unsynced
-		_, cancel := w.Watch(testKey, true, 1)
-		cancels[i] = cancel
+		watchID := w.Watch(testKey, true, 1)
+		watchIDs[i] = watchID
```
Comment (heyitsanthony) [inline storage/watchable_store_test.go]: `watchIDs[i] = w.Watch(...)`

File: storage/watchable_store_bench_test.go (hunk 5 above)
Comment (gyuho) [inline storage/watchable_store_bench_test.go]: Ok will change those. Thanks,

File: storage/watcher.go [hunk 7]
```diff
@@ -98,6 +97,7 @@ func (ws *watchStream) Cancel(id int64) error {
 	}
 	cancel()
 	delete(ws.cancels, id)
+
```
Comment (heyitsanthony) [inline storage/watcher.go]: unnecessary new blank line

File: storage/watcher.go [hunk 8]
```diff
@@ -108,6 +108,7 @@ func (ws *watchStream) Close() {
 	for _, cancel := range ws.cancels {
 		cancel()
 	}
+
```
Comment (heyitsanthony) [inline storage/watcher.go]: unnecessary new blank line

File: storage/watcher_test.go [hunk 9]
```diff
@@ -93,8 +97,4 @@ func TestWatchStreamCancelWatcherByID(t *testing.T) {
 			t.Errorf("#%d: err = %v, want %v", i, gerr, tt.werr)
 		}
 	}
-
-	if l := len(w.(*watchStream).cancels); l != 0 {
```
Comment (heyitsanthony) [inline storage/watcher_test.go]: any reason for removing this?

Comment (gyuho) [inline storage/watcher_test.go]: Actually, I shouldn't have deleted it. Will put it back. Thanks

**Comment by gyuho** [comment]
@heyitsanthony @xiang90 Just fixed some based on previous feedback. PTAL.

**Comment by xiang90** [comment]
LGTM

**Comment by gyuho** [comment]
Thanks. Will merge after tests and https://github.com/coreos/etcd/pull/4131 merge.

**Comment by heyitsanthony** [comment]
lgtm after squashing

**Comment by gyuho** [comment]
Ok, I will squash commits into one.

**Comment by gyuho** [comment]
Will merge after CI passes.

**Pull request merged**